src/main/java/org/threadly/concurrent/PriorityScheduler.java

```diff
@@ -763,232 +639,543 @@ protected void addToExecuteQueue(OneTimeTaskWrapper task) {
         }
       }
     }
-    
-    if (w != null) {  // may be null if shutdown
-      w.nextTask(task);
+
+    /**
+     * Change the set idle worker keep alive time.  If this is a reduction in the previously set 
+     * keep alive time, this call will then check for expired workers.
+     * 
+     * @param keepAliveTimeInMs New keep alive time in milliseconds
+     */
+    public void setKeepAliveTime(long keepAliveTimeInMs) {
+      ArgumentVerifier.assertNotNegative(keepAliveTimeInMs, "keepAliveTimeInMs");
+      
+      boolean checkForExpiredWorkers = this.keepAliveTimeInMs > keepAliveTimeInMs;
+      
+      this.keepAliveTimeInMs = keepAliveTimeInMs;
+      
+      if (checkForExpiredWorkers) {
+        synchronized (workersLock) {
+          expireOldWorkers();
+        }
+      }
     }
-  }
 
-  /**
-   * After a task has been pulled from the queue and is ready to execute it is provided here.  
-   * This function will get an available worker, waiting a bit of time for one to become 
-   * available if none are immediately available.  If after that there is still none available it 
-   * will create one (assuming we have not reached our max pool size).  Then the acquired worker 
-   * will be provided the task to execute.
-   * 
-   * @param task Task to execute once we have an available worker
-   * @throws InterruptedException Thrown if thread is interrupted while waiting for a worker
-   */
-  protected void runLowPriorityTask(TaskWrapper task) throws InterruptedException {
-    Worker w = null;
-    synchronized (workersLock) {
-      if (! shutdownFinishing) {
-        // wait for high priority tasks that have been waiting longer than us if all workers are consumed
-        long waitAmount;
-        while (currentPoolSize >= maxPoolSize && 
-               availableWorkers.size() < WORKER_CONTENTION_LEVEL &&   // only care if there is worker contention
-               ! shutdownFinishing &&
-               ! highPriorityConsumer.isQueueEmpty() && 
-               (waitAmount = task.getDelayEstimateInMillis() - lastHighDelay) > LOW_PRIORITY_WAIT_TOLLERANCE_IN_MS) {
-          workersLock.wait(waitAmount);
-          Clock.systemNanoTime(); // update for getDelayEstimateInMillis
+    /**
+     * Changes the setting weather core threads are allowed to be killed if they remain idle.  If 
+     * changing to allow core thread timeout, this call will then perform a check to look for 
+     * expired workers.
+     * 
+     * @param value {@code true} if core threads should be expired when idle.
+     */
+    public void allowCoreThreadTimeOut(boolean value) {
+      boolean checkForExpiredWorkers = ! allowCorePoolTimeout && value;
+      
+      allowCorePoolTimeout = value;
+      
+      if (checkForExpiredWorkers) {
+        synchronized (workersLock) {
+          expireOldWorkers();
         }
-        // check if we should reset the high delay for future low priority tasks
-        if (highPriorityConsumer.isQueueEmpty()) {
```

Comment:
We had talked in person about how this logic has changed...since we don't have a handle on the queue any more, it's the queue now that must notify the WorkerPool when there is no remaining high priority tasks (I will put a comment where that is happening).